File: Python/TugOfWar.py
'''
r=range,d=[]
n=int(input())
for i in r(n): d.append(int(input()))
x=[[0 for i in r(45000)] for y in r(50)]
x[0][0]=1
for z in d:
 for i,j in zip(r(50),r(45000)):
  if x[j][i]:x[j+1][i+z]:
   pass
'''
###

n = int (input())
x=[[0 for i in range(30)] for y in range(3)]
x[0][0] = 1
t=0
for i in range(n):
 d=int(input())
 t+=d
 for j in range(n//2)[::-1]:
  for k in range (30-d):
   x[j+1][k+d] += x[j][k]
a=t//2+1;i=0
print(x)
while 1:
 if x[n//2][a+i]: print (a-i, t-a+i);break
 elif x[n//2][a-i]: print (a+i, t-a-i);break
 else: i+=1

File: C/TugOfWar.c
#include<stdio.h>

int arr[105];
int dp [102][46000];

int ABS(int x){
    if (x < 0) return -x;
    else return x;
}

int main(){
    int n;
    int i, j, k;

    // Scan in input
    scanf("%d", &n);
    for (i=0; i<n; ++i){
        scanf("%d", &arr[i]);
    }

    // Initialise my dp array
    dp[0][0] = 1;
    int total = 0;

    // Populate the dp array
    // Each element represents whether it's possible to be in that state
    for (i=0; i<n; ++i){
        for (j=i/2; j>=0; --j){
            for (k=45000-arr[i]; k>=0; --k){
                dp[j+1][k+arr[i]] += dp[j][k];
            }
        }
        total += arr[i];
    }

    // Find the best
    int currbest = 0;
    i = 0;
    while (i < total){
        if (dp[n/2][i]){
            if (ABS((total-2*i)) < ABS((total-2*currbest))){
                currbest = i;
            }
            else {
                break;
            }
        }
        ++i;
    }

    if (currbest < total - currbest){
        currbest = total - currbest;
    }

    printf ("%d %d\n", total-currbest, currbest);

    return 0;
}
